ratis-shell/src/main/java/org/apache/ratis/shell/cli/sh/command/AbstractRatisCommand.java

```diff
@@ -0,0 +1,179 @@
+    List<RaftPeer> peers = addresses.stream()
+        .map(addr -> RaftPeer.newBuilder()
+            .setId(RaftUtils.getPeerId(addr))
+            .setAddress(addr)
+            .build()
+        ).collect(Collectors.toList());
+    raftGroup = RaftGroup.valueOf(raftGroupIdFromConfig, peers);
+    try (final RaftClient client = RaftUtils.createClient(raftGroup)) {
+      RaftGroupId remoteGroupId;
+      List<RaftGroupId> groupIds;
+      groupIds = RetryUtil.run(peers,
+          p -> {
+            try {
+              return client.getGroupManagementApi((p.getId())).list().getGroupIds();
+            } catch (IOException e) {
+              e.printStackTrace();
+              return null;
+            }
+          }
+      );
+
+      if (groupIds.size() == 1) {
+        remoteGroupId = groupIds.get(0);
+      } else {
+        final UUID raftGroupUuid = raftGroupIdFromConfig.getUuid();
+        Optional<RaftGroupId> raftGroupId =
+            groupIds.stream().filter(r -> raftGroupUuid.equals(r.getUuid()))
+                .findFirst();
+        if (!raftGroupId.isPresent()) {
+          printStream.println(
+              "there are more than one group, you should specific one."
+                  + groupIds);
+          return -1;
+        } else {
+          remoteGroupId = raftGroupId.get();
+        }
+      }
+      raftGroup = RetryUtil.run(peers,
+          p -> {
+            try {
+              return client.getGroupManagementApi((p.getId()))
+                  .info(remoteGroupId).getGroup();
+            } catch (IOException e) {
+              e.printStackTrace();
+              return null;
+            }
+          }
+      );
+    }
+    return 0;
+  }
+
+  @Override
+  public void validateArgs(CommandLine cl) throws IllegalArgumentException {
+    if (!cl.hasOption(PEER_OPTION_NAME)) {
+      throw new IllegalArgumentException(String.format(
+          "should provide [%s]", PEER_OPTION_NAME));
+    }
+  }
+
+  @Override
+  public Options getOptions() {
+    return new Options()
+            .addOption(PEER_OPTION_NAME, true, "Peer addresses seperated by comma")
+            .addOption(GROUPID_OPTION_NAME, true, "Raft group id");
+  }
+
+  protected PrintStream getPrintStream() {
+    return printStream;
+  }
+
+  protected RaftGroup getRaftGroup() {
+    return raftGroup;
+  }
+
+  /**
+   * Get the leader id.
+   *
+   * @param roleInfo the role info
+   * @return the leader id
+   */
+  protected RaftPeerProto getLeader(RoleInfoProto roleInfo) {
+    if (roleInfo == null) {
+      return null;
+    }
+    if (roleInfo.getRole() == RaftPeerRole.LEADER) {
+      return roleInfo.getSelf();
+    }
+    FollowerInfoProto followerInfo = roleInfo.getFollowerInfo();
+    if (followerInfo == null) {
+      return null;
+    }
+    return followerInfo.getLeaderInfo().getId();
+  }
+
+  protected void processReply(RaftClientReply reply, String msg)
+      throws IOException {
+    RaftUtils.processReply(reply, msg, printStream);
```

Comment:
Just put the code from RaftUtils.processReply here.

Reply:
OK